### PR TITLE
Add trainer for models estimating distances in a Cayley graph

### PR DESCRIPTION
**Pull request by stasdiener**
Adds `cayleypy/train/config.py` and `cayleypy/train/trainer.py`: the core of a trainer for models that estimate distance from the central state.

## What it does

`TrainConfig` (frozen dataclass) describes the whole process: how much data to generate (`n_walks`, `rw_length`, `rw_mode`, `nbt_history_depth`), the optimizer (`lr`, `lr_min`, `weight_decay`), weight averaging (`ema_decay`), the loss (`loss`, `tau`), plus `seed` and `verbose`. All values are validated on construction, so a misconfigured run fails immediately rather than after an epoch of data generation.

`Trainer` runs it:

- **data** — on every epoch a fresh set of random walks is generated with `RandomWalksGenerator` (default mode is `nbt`, which mixes fastest, so the number of steps is the closest estimate of the true distance). Because data is regenerated, the loss per epoch reads as a validation curve.
- **optimization** — AdamW + `CosineAnnealingLR` from `lr` to `lr_min` over `n_epochs`.
- **weight averaging** — an EMA copy of the weights is maintained and is what `predictor()` and `save()` use by default. It is also the frozen target copy that Bellman/DAVI fine-tuning will need later.
- **checkpoints** — `save()` writes a self-describing checkpoint (config + weights + hash of the graph), `Trainer.from_checkpoint()` continues training from one. Loading a checkpoint for another graph fails instead of silently producing nonsense.
- **progress** — `print` when `TrainConfig.verbose >= 1`, no progress-bar dependency.

`train_step(states, targets, mask=..., weights=...)` is public, so the data sources and the Bellman mode that come next can feed their own batches (including sparsely labeled ones) into the same loop.

Scope of this PR is deliberately the core loop only: single-output models trained on random walks. BFS anchors, sparse-Q labels and beam-search paths are the next PR; a Q-model (`n_outputs != 1`) is rejected with a clear message until then, because random walks do not provide a target for every output.

## Tests

20 new tests in `cayleypy/train/trainer_test.py`. Fast ones cover mechanics: number of optimizer steps and losses returned, shape/dtype/values of generated data, determinism under a seed, one step reducing the loss on a fixed batch, a fully masked batch leaving the weights untouched, the EMA relation checked against `decay*ema + (1-decay)*w` on two consecutive steps, EMA disabled, the cosine schedule ending at `lr_min`, `predictor()`/`save()` using averaged weights (and both opting out), checkpoint round trip and resume, `verbose` output, and the error cases (every invalid config field, model for states of another size, multi-output model, too few one-hot classes, empty data, checkpoint for another graph).

Two slow tests (`RUN_SLOW_TESTS=1`) on `lrx(5)`, seeded:

- with exact labels (`rw_mode="bfs"` visits the whole graph on such a small graph) the final loss is under 0.2 of the initial one, mean absolute error against exact BFS is under 0.5, and beam search with the trained model finds an **optimal** path for all 120 states with **beam width 1** — where the default Hamming heuristic solves 10.
- trained on non-backtracking random walks, the model solves **all 120 states with beam width 5** — where the Hamming heuristic solves 23. These targets are overestimates, so the loss plateaus at the label noise (about 0.4 of the initial loss) and the model does not predict true distances; it is still a much better beam-search heuristic.

## Relation to upstream cayleypy/cayleypy#151

[cayleypy/cayleypy#151](https://github.com/cayleypy/cayleypy/pull/151) by @vlzm proposed a trainer (in `cayleypy/trainers/`) and has been approved but unmerged since October 2025; the pipeline here is the same idea, and credit for it goes to that work. Differences:

- built on the `score_children` contract and `ModelConfig`/checkpoint format from the contract PR, so trained models are usable by beam search and reloadable without knowing the architecture;
- losses come from the separate losses PR, including masked-sparse, which is what makes the upcoming sparse-Q labels possible;
- EMA of the weights;
- zero new dependencies (#151 added some to `pyproject.toml`) — progress is a `verbose` flag and `print`, as in `beam_search.py`;
- lives in `cayleypy/train/`, so it does not conflict with `cayleypy/trainers/` from #151.

## Stacking

This PR depends on both the contract PR (checkpoints) and the losses PR, which are not merged yet. `base/trainer-core` is a mechanical merge of those two branches, used as the base so that the diff here shows only the trainer. Draft until they land.

## Checks

`./lint.sh` clean (black 68 files, pylint 10.00/10, mypy 68 files), `black --check .` clean over the whole repo, `docs/build_docs.sh` (`-W`) clean, doctests pass. `RUN_SLOW_TESTS=1 pytest`: 366 passed / 12 skipped / 3 xfailed on Python 3.12 (torch 2.13) and on Python 3.9 (torch 2.8).

---

Based on `base/trainer-core`, a mechanical merge of #191 and #192 with no changes of its own. Both should be merged first.

### CODE REVIEW

File: cayleypy/train/trainer.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+        :param states: States (in decoded representation) to train on.
+        :param targets: Target distances for `states`.
+        :param mask: Which targets are known (optional), see :class:`cayleypy.train.Loss`.
+        :param weights: Importance of every target (optional), see :class:`cayleypy.train.Loss`.
+        :return: Value of the loss on this batch, before the step.
+        """
+        self.model.train()
+        self.optimizer.zero_grad(set_to_none=True)
+        loss = self.loss(self.model(states), targets, mask=mask, weights=weights)
```
Comment (Alicentpt) [inline cayleypy/train/trainer.py:197]: here if mask is fully filled (i.e. no labeled data), we still perform backward calculation, optimizer step and ema update, what may change some model and optimizer states, by doing nothing at the step
